# Configuration de la connexion à la base de données  

In [1]:
import pandas as pd 
import cx_Oracle
import sys
import os

%pylab inline
%load_ext sql

%pylab is deprecated, use %matplotlib inline and import the required libraries.
Populating the interactive namespace from numpy and matplotlib


In [2]:
sys.path.append(os.path.abspath('../../.'))
from dbconnect import utilisateur,connexion,password

# Connexion 

In [3]:
%sql oracle+cx_oracle://$utilisateur:$password@$connexion

<img src="https://raw.githubusercontent.com/rbizoi/IntelligenceEnDonneesDeSante/main/InterrogationSQL/images/612.png" width="800">

# L’opérateur UNION

<img src="https://raw.githubusercontent.com/rbizoi/IntelligenceEnDonneesDeSante/main/InterrogationSQL/images/union.png" width="400">


In [4]:
%%sql 
SELECT SOCIETE,VILLE,'Client' FROM CLIENTS 
UNION 
SELECT SOCIETE,VILLE,'Fournisseur' FROM FOURNISSEURS 

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Alfreds Futterkiste', 'Berlin', 'Client'),
 ('Ana Trujillo Emparedados y helados', 'México D.F.', 'Client'),
 ('Antonio Moreno Taquería', 'México D.F.', 'Client'),
 ('Around the Horn', 'London', 'Client'),
 ('Aux joyeux ecclésiastiques', 'Paris', 'Fournisseur'),
 ("B's Beverages", 'London', 'Client'),
 ('Berglunds snabbköp', 'Luleå', 'Client'),
 ('Bigfoot Breweries', 'Bend', 'Fournisseur'),
 ('Blauer See Delikatessen', 'Mannheim', 'Client'),
 ('Blondel père et fils', 'Strasbourg', 'Client'),
 ("Bon app'", 'Marseille', 'Client'),
 ('Bottom-Dollar Markets', 'Tsawassen', 'Client'),
 ('Bólido Comidas preparadas', 'Madrid', 'Client'),
 ('Cactus Comidas para llevar', 'Buenos Aires', 'Client'),
 ('Centro comercial Moctezuma', 'México D.F.', 'Client'),
 ('Chop-suey Chinese', 'Bern', 'Client'),
 ('Comércio Mineiro', 'São Paulo', 'Client'),
 ('Consolidated Holdings', 'London', 'Client'),
 ("Cooperativa de Quesos 'Las Cabras'", 'Oviedo', 'Fournisseur'),
 ('Die Wandernde Kuh', 'Stuttgart', 'Client'),
 ('Drachenblut Delikatessen', 'Aachen', 'Client'),
 ('Du monde entier', 'Nantes', 'Client'),
 ('Eastern Connection', 'London', 'Client'),
 ('Ernst Handel', 'Graz', 'Client'),
 ('Escargots Nouveaux', 'Montceau', 'Fournisseur'),
 ('Exotic Liquids', 'London', 'Fournisseur'),
 ('FISSA Fabrica Inter. Salchichas S.A.', 'Madrid', 'Client'),
 ('Familia Arquibaldo', 'São Paulo', 'Client'),
 ('Folies gourmandes', 'Lille', 'Client'),
 ('Folk och fä HB', 'Bräcke', 'Client'),
 ('Formaggi Fortini s.r.l.', 'Ravenna', 'Fournisseur'),
 ("Forêts d'érables", 'Ste-Hyacinthe', 'Fournisseur'),
 ('France restauration', 'Nantes', 'Client'),
 ('Franchi S.p.A.', 'Torino', 'Client'),
 ('Frankenversand', 'München', 'Client'),
 ('Furia Bacalhau e Frutos do Mar', 'Lisboa', 'Client'),
 ("G'day, Mate", 'Sydney', 'Fournisseur'),
 ('GROSELLA-Restaurante', 'Caracas', 'Client'),
 ('Gai pâturage', 'Annecy', 'Fournisseur'),
 ('Galería del gastrónomo', 'Barcelona', 'Client'),
 ('Godos Cocina Típica', 'Sevilla', 'Client'),
 ('Gourmet Lanchonetes', 'Campinas', 'Client'),
 ("Grandma Kelly's Homestead", 'Ann Arbor', 'Fournisseur'),
 ('Great Lakes Food Market', 'Eugene', 'Client'),
 ('HILARIÓN-Abastos', 'San Cristóbal', 'Client'),
 ('Hanari Carnes', 'Rio de Janeiro', 'Client'),
 ('Heli Süßwaren GmbH  Co. KG', 'Berlin', 'Fournisseur'),
 ('Hungry Coyote Import Store', 'Elgin', 'Client'),
 ('Hungry Owl All-Night Grocers', 'Cork', 'Client'),
 ('Island Trading', 'Cowes', 'Client'),
 ('Karkki Oy', 'Lappeenranta', 'Fournisseur'),
 ('Kelly', 'Ann Arbor', 'Fournisseur'),
 ('Königlich Essen', 'Brandenburg', 'Client'),
 ('LILA-Supermercado', 'Barquisimeto', 'Client'),
 ('LINO-Delicateses', 'I. de Margarita', 'Client'),
 ("La corne d'abondance", 'Versailles', 'Client'),
 ("La maison d'Asie", 'Toulouse', 'Client'),
 ('Laughing Bacchus Wine Cellars', 'Vancouver', 'Client'),
 ('Lazy K Kountry Store', 'Walla Walla', 'Client'),
 ('Lehmanns Marktstand', 'Frankfurt a.M.', 'Client'),
 ('Leka Trading', 'Singapore', 'Fournisseur'),
 ("Let's Stop N Shop", 'San Francisco', 'Client'),
 ('Lonesome Pine Restaurant', 'Portland', 'Client'),
 ('Lyngbysild', 'Lyngby', 'Fournisseur'),
 ('Ma Maison', 'Montréal', 'Fournisseur'),
 ('Magazzini Alimentari Riuniti', 'Bergamo', 'Client'),
 ('Maison Dewey', 'Bruxelles', 'Client'),
 ("Mayumi's", 'Osaka', 'Fournisseur'),
 ('Morgenstern Gesundkost', 'Leipzig', 'Client'),
 ('Mère Paillarde', 'Montréal', 'Client'),
 ('New England Seafood Cannery', 'Boston', 'Fournisseur'),
 ('Nord-Ost-Fisch Handelsgesellschaft mbH', 'Cuxhaven', 'Fournisseur'),
 ('Norske Meierier', 'Sandvika', 'Fournisseur'),
 ('North/South', 'London', 'Client'),
 ('Nouvelle-Orléans Cajun Delights', 'Nouvelle-Orléans', 'Fournisseur'),
 ('Océano Atlántico Ltda.', 'Buenos Aires', 'Client'),
 ('Old World Delicatessen', 'Anchorage', 'Client'),
 ('Ottilies Käseladen', 'Köln', 'Client'),
 ('PB Knäckebröd AB', 'Göteborg', 'Fournisseur'),
 ('Paris spécialités', 'Paris', 'Client'),
 ('Pasta Buttini s.r.l.', 'Salerno', '

# L’opérateur INTERSECT

<img src="https://raw.githubusercontent.com/rbizoi/IntelligenceEnDonneesDeSante/main/InterrogationSQL/images/intersect.png" width="400">


In [5]:
%%sql 
SELECT VILLE,                                        
       NOM_PRODUIT                                   
FROM CLIENTS CL                                      
     JOIN COMMANDES CO                               
        ON ( CL.CODE_CLIENT = CO.CODE_CLIENT)        
     JOIN DETAILS_COMMANDES DC                       
        ON ( CO.NO_COMMANDE = DC.NO_COMMANDE)        
     JOIN PRODUITS PR                                
        ON ( DC.REF_PRODUIT = PR.REF_PRODUIT)        
INTERSECT                                            
SELECT VILLE, NOM_PRODUIT                            
FROM PRODUITS PR                                     
     JOIN FOURNISSEURS FR                            
        ON ( PR.NO_FOURNISSEUR = FR.NO_FOURNISSEUR )

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Berlin', 'Granola'),
 ('Berlin', 'Gumbär Gummibärchen'),
 ('Berlin', 'NuNuCa Nuß-Nougat-Creme'),
 ('Berlin', 'Schoggi Schokolade'),
 ('Frankfurt a.M.', 'Original Frankfurter grüne Soße'),
 ('Frankfurt a.M.', 'Rhönbräu Klosterbier'),
 ('Frankfurt a.M.', 'Rössle Sauerkraut'),
 ('Frankfurt a.M.', 'Thüringer Rostbratwurst'),
 ('Frankfurt a.M.', 'Wimmers gute Semmelknödel'),
 ('London', 'Aniseed Syrup'),
 ('London', 'Chai'),
 ('London', 'Chang'),
 ('Montréal', 'Pears'),
 ('Montréal', 'Pâté chinois'),
 ('Montréal', 'Tourtière'),
 ('Paris', 'Beer'),
 ('Paris', 'Chartreuse verte'),
 ('Paris', 'Côte de Blaye'),
 ('Paris', 'Hot Cereal'),
 ('Paris', 'Mozzarella'),
 ('São Paulo', 'Guaraná Fantástica')]

# L’opérateur DIFFERENCE

<img src="https://raw.githubusercontent.com/rbizoi/IntelligenceEnDonneesDeSante/main/InterrogationSQL/images/difference.png" width="400">


In [6]:
%%sql 
SELECT CO.NO_COMMANDE,                          
       CO.DATE_COMMANDE                         
FROM CLIENTS CL                                 
     JOIN COMMANDES CO                          
        ON ( CL.CODE_CLIENT = CO.CODE_CLIENT)   
     JOIN DETAILS_COMMANDES DC                  
        ON ( CO.NO_COMMANDE = DC.NO_COMMANDE)   
     JOIN PRODUITS PR                           
        ON ( DC.REF_PRODUIT = PR.REF_PRODUIT)   
WHERE PR.CODE_CATEGORIE = 1                     
MINUS                                           
SELECT CO.NO_COMMANDE,                          
    CO.DATE_COMMANDE                            
FROM CLIENTS CL                                 
     JOIN COMMANDES CO                          
        ON ( CL.CODE_CLIENT = CO.CODE_CLIENT)   
     JOIN DETAILS_COMMANDES DC                  
        ON ( CO.NO_COMMANDE = DC.NO_COMMANDE)   
     JOIN PRODUITS PR                           
        ON ( DC.REF_PRODUIT = PR.REF_PRODUIT)   
WHERE PR.CODE_CATEGORIE = 9                     

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[(215025, datetime.datetime(2010, 1, 3, 0, 0)),
 (215130, datetime.datetime(2010, 1, 9, 0, 0)),
 (215220, datetime.datetime(2010, 1, 13, 0, 0)),
 (215226, datetime.datetime(2010, 1, 13, 0, 0)),
 (215230, datetime.datetime(2010, 1, 13, 0, 0)),
 (215436, datetime.datetime(2010, 1, 22, 0, 0)),
 (215441, datetime.datetime(2010, 1, 22, 0, 0)),
 (215444, datetime.datetime(2010, 1, 22, 0, 0)),
 (215472, datetime.datetime(2010, 1, 24, 0, 0)),
 (215483, datetime.datetime(2010, 1, 24, 0, 0)),
 (215496, datetime.datetime(2010, 1, 25, 0, 0)),
 (215505, datetime.datetime(2010, 1, 26, 0, 0)),
 (215508, datetime.datetime(2010, 1, 26, 0, 0)),
 (215513, datetime.datetime(2010, 1, 26, 0, 0)),
 (215536, datetime.datetime(2010, 1, 27, 0, 0)),
 (215589, datetime.datetime(2010, 1, 30, 0, 0)),
 (215591, datetime.datetime(2010, 1, 30, 0, 0)),
 (215598, datetime.datetime(2010, 1, 30, 0, 0)),
 (215603, datetime.datetime(2010, 1, 30, 0, 0)),
 (215604, datetime.datetime(2010, 1, 30, 0, 0)),
 (215624, datetime.datetime(2010, 1, 31, 0, 0)),
 (215876, datetime.datetime(2010, 2, 12, 0, 0)),
 (215884, datetime.datetime(2010, 2, 12, 0, 0)),
 (215888, datetime.datetime(2010, 2, 12, 0, 0)),
 (215894, datetime.datetime(2010, 2, 12, 0, 0)),
 (215904, datetime.datetime(2010, 2, 13, 0, 0)),
 (215982, datetime.datetime(2010, 2, 17, 0, 0)),
 (216029, datetime.datetime(2010, 2, 19, 0, 0)),
 (216033, datetime.datetime(2010, 2, 19, 0, 0)),
 (216043, datetime.datetime(2010, 2, 19, 0, 0)),
 (216091, datetime.datetime(2010, 2, 22, 0, 0)),
 (216092, datetime.datetime(2010, 2, 22, 0, 0)),
 (216093, datetime.datetime(2010, 2, 22, 0, 0)),
 (216095, datetime.datetime(2010, 2, 22, 0, 0)),
 (216096, datetime.datetime(2010, 2, 22, 0, 0)),
 (216098, datetime.datetime(2010, 2, 22, 0, 0)),
 (216102, datetime.datetime(2010, 2, 22, 0, 0)),
 (216103, datetime.datetime(2010, 2, 22, 0, 0)),
 (216121, datetime.datetime(2010, 2, 23, 0, 0)),
 (216129, datetime.datetime(2010, 2, 23, 0, 0)),
 (216131, datetime.datetime(2010, 2, 23, 0, 0)),
 (216140, datetime.datetime(2010, 2, 23, 0, 0)),
 (216142, datetime.datetime(2010, 2, 23, 0, 0)),
 (216205, datetime.datetime(2010, 2, 26, 0, 0)),
 (216209, datetime.datetime(2010, 2, 26, 0, 0)),
 (216212, datetime.datetime(2010, 2, 26, 0, 0)),
 (216214, datetime.datetime(2010, 2, 26, 0, 0)),
 (216242, datetime.datetime(2010, 2, 28, 0, 0)),
 (216333, datetime.datetime(2010, 3, 5, 0, 0)),
 (216334, datetime.datetime(2010, 3, 5, 0, 0)),
 (216336, datetime.datetime(2010, 3, 5, 0, 0)),
 (216338, datetime.datetime(2010, 3, 5, 0, 0)),
 (216355, datetime.datetime(2010, 3, 6, 0, 0)),
 (216358, datetime.datetime(2010, 3, 6, 0, 0)),
 (216365, datetime.datetime(2010, 3, 6, 0, 0)),
 (216368, datetime.datetime(2010, 3, 6, 0, 0)),
 (216430, datetime.datetime(2010, 3, 10, 0, 0)),
 (216439, datetime.datetime(2010, 3, 10, 0, 0)),
 (216448, datetime.datetime(2010, 3, 10, 0, 0)),
 (216449, datetime.datetime(2010, 3, 10, 0, 0)),
 (216453, datetime.datetime(2010, 3, 11, 0, 0)),
 (216464, datetime.datetime(2010, 3, 11, 0, 0)),
 (216472, datetime.datetime(2010, 3, 11, 0, 0)),
 (216516, datetime.datetime(2010, 3, 14, 0, 0)),
 (216518, datetime.datetime(2010, 3, 14, 0, 0)),
 (216534, datetime.datetime(2010, 3, 14, 0, 0)),
 (216581, datetime.datetime(2010, 3, 17, 0, 0)),
 (216585, datetime.datetime(2010, 3, 17, 0, 0)),
 (216587, datetime.datetime(2010, 3, 17, 0, 0)),
 (216592, datetime.datetime(2010, 3, 17, 0, 0)),
 (216594, datetime.datetime(2010, 3, 17, 0, 0)),
 (216602, datetime.datetime(2010, 3, 17, 0, 0)),
 (216608, datetime.datetime(2010, 3, 18, 0, 0)),
 (216610, datetime.datetime(2010, 3, 18, 0, 0)),
 (216614, datetime.datetime(2010, 3, 18, 0, 0)),
 (216628, datetime.datetime(2010, 3, 19, 0, 0)),
 (216648, datetime.datetime(2010, 3, 19, 0, 0)),
 (216650, datetime.datetime(2010, 3, 19, 0, 0)),
 (216659, datetime.datetime(2010, 3, 20, 0, 0)),
 (216664, datetime.datetime(2010, 3, 20, 0, 0)),
 (216668, datetime.datetime(2010, 3, 20, 0, 0)),
 (216670, datetime.datetime(2010, 3, 20, 

# Sous-requête monolignes

<img src="https://raw.githubusercontent.com/rbizoi/IntelligenceEnDonneesDeSante/main/InterrogationSQL/images/sr01.png" width="400">


In [7]:
%%sql 
SELECT NOM_PRODUIT 
FROM PRODUITS 
WHERE UNITES_STOCK = (SELECT MAX(UNITES_STOCK) FROM PRODUITS)

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Potato Chips',)]

In [8]:
%%sql 
SELECT CL.SOCIETE,                                     
       CO.NO_COMMANDE,                                 
       COUNT( DC.REF_PRODUIT)                          
FROM CLIENTS CL                                        
     JOIN COMMANDES CO                                 
        ON ( CL.CODE_CLIENT = CO.CODE_CLIENT)          
     JOIN DETAILS_COMMANDES DC                         
        ON ( CO.NO_COMMANDE = DC.NO_COMMANDE)          
GROUP BY CL.SOCIETE,                                   
         CO.NO_COMMANDE                                
HAVING COUNT( DC.REF_PRODUIT) >=                       
   ( SELECT COUNT( DC.REF_PRODUIT)                     
     FROM COMMANDES CO                                 
          JOIN DETAILS_COMMANDES DC                    
               ON ( CO.NO_COMMANDE = DC.NO_COMMANDE)   
     WHERE  CO.NO_COMMANDE = 224975)

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Hungry Owl All-Night Grocers', 220267, 52),
 ('Rancho grande', 215629, 52),
 ('Berglunds snabbköp', 216798, 52),
 ('Blondel père et fils', 224968, 52),
 ('Furia Bacalhau e Frutos do Mar', 225524, 52),
 ('Rattlesnake Canyon Grocery', 228419, 52),
 ('LINO-Delicateses', 224963, 52),
 ('Du monde entier', 216814, 52),
 ('White Clover Markets', 220279, 52),
 ('Laughing Bacchus Wine Cellars', 224974, 52),
 ('Die Wandernde Kuh', 216810, 52),
 ('The Big Cheese', 224975, 52)]

# Sous-requête multilignes

<img src="https://raw.githubusercontent.com/rbizoi/IntelligenceEnDonneesDeSante/main/InterrogationSQL/images/sr02.png" width="400">


In [9]:
%%sql 
SELECT SOCIETE,                                                
       VILLE,                                                  
       PAYS                                                    
FROM CLIENTS                                                   
WHERE  CODE_CLIENT IN (                                        
               SELECT CO.CODE_CLIENT                           
               FROM COMMANDES CO                               
                    JOIN EMPLOYES EM                           
                         ON ( EM.NO_EMPLOYE = CO.NO_EMPLOYE)   
               WHERE EM.REND_COMPTE = 11)

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Wolski  Zajazd', 'Warszawa', 'Pologne'),
 ('Chop-suey Chinese', 'Bern', 'Suisse'),
 ('Ernst Handel', 'Graz', 'Autriche'),
 ('Toms Spezialitäten', 'Münster', 'Allemagne'),
 ('Lehmanns Marktstand', 'Frankfurt a.M.', 'Allemagne'),
 ('Alfreds Futterkiste', 'Berlin', 'Allemagne'),
 ('Blauer See Delikatessen', 'Mannheim', 'Allemagne'),
 ('Königlich Essen', 'Brandenburg', 'Allemagne'),
 ('Richter Supermarkt', 'Genève', 'Suisse'),
 ('Ottilies Käseladen', 'Köln', 'Allemagne'),
 ('Morgenstern Gesundkost', 'Leipzig', 'Allemagne'),
 ('QUICK-Stop', 'Cunewalde', 'Allemagne'),
 ('Frankenversand', 'München', 'Allemagne'),
 ('Piccolo und mehr', 'Salzburg', 'Autriche'),
 ('Drachenblut Delikatessen', 'Aachen', 'Allemagne'),
 ('Die Wandernde Kuh', 'Stuttgart', 'Allemagne')]

# Sous-requête synchronisée

<img src="https://raw.githubusercontent.com/rbizoi/IntelligenceEnDonneesDeSante/main/InterrogationSQL/images/sr03.png" width="400">


In [10]:
%%sql 
SELECT 
    CODE_CATEGORIE "Cat",                           
    NOM_PRODUIT,                                    
    UNITES_STOCK  "Stock",                          
    PRIX_UNITAIRE "Prix"                            
FROM PRODUITS P1                                       
WHERE UNITES_STOCK >(                                  
         SELECT AVG(UNITES_STOCK)*2                    
         FROM PRODUITS P2                              
         WHERE P2.CODE_CATEGORIE = P1.CODE_CATEGORIE)

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[(2, "Grandma's Boysenberry Spread", 120, Decimal('125')),
 (4, 'Queso Manchego La Pastora', 86, Decimal('190')),
 (3, 'NuNuCa Nuß-Nougat-Creme', 76, Decimal('70')),
 (8, 'Boston Crab Meat', 123, Decimal('92')),
 (4, 'Raclette Courdavault', 79, Decimal('275')),
 (4, 'Geitost', 112, Decimal('13')),
 (3, 'Valkoinen suklaa', 65, Decimal('81')),
 (2, "Sirop d'érable", 113, Decimal('143')),
 (1, 'Rhönbräu Klosterbier', 125, Decimal('39')),
 (2, 'Syrup', 100, Decimal('7.5')),
 (9, 'Boysenberry Spread', 100, Decimal('18.75')),
 (1, 'Green Tea', 125, Decimal('2')),
 (7, 'Potato Chips', 200, Decimal('0.49'))]

# Atelier 8.2	Les jointures complexes
## Les sous-requêtes monolignes
Écrivez les requêtes vous permettant d’afficher :

>> * Les produits pour lesquels la quantité en stock est inférieure à la moyenne.
>> * Les sociétés clientes et les numéros des commandes qui ont un nombre égal ou supérieur de produits achetés que la commande numéro '216798'.


## Les sous-requêtes multilignes
Écrivez les requêtes vous permettant d’afficher :
>> * Les sociétés clientes et leurs commandes pour tous les produits livrés par un fournisseur qui habite Paris.
>> * Les sociétés clientes françaises qui ont commandé le produit ‘Chai’ mais également qui ont commandé plus de 39 produits le '01/05/2011'.

## Les sous-requêtes synchronisée
Écrivez les requêtes vous permettant d’afficher :
>> * Les clients pour lesquels les frais de ports par commande dépassent leur moyenne globale des frais de ports.
>> * Le produit, le fournisseur et les unités en stock pour les produits qui ont un stock inférieur à la moyenne des unités en stock pour les produits du même fournisseur.
>> * Les clients et ses commandes pour les clients qui payent un port supérieur à la moyenne des commandes pour la même année.
>> * Les employés avec leur salaire et le pourcentage correspondant par rapport au total de la masse salariale par fonction. Essayez d’utiliser une sous-requête dans la clause « FROM ».
